### ISSUE TITLE
SSLContext.wrap_socket sends SNI Extension when server_hostname is IP

### ISSUE THREAD

**Issue by efccb2b9-c9f3-44b4-8086-08881f25a009** [python/cpython#76366]
BPO | [32185](https://bugs.python.org/issue32185)
--- | :---
Nosy | @pfmoore, @pitrou, @tiran, @tjguk, @alex, @zware, @zooba, @dstufft, @nitzmahone
PRs | <li>python/cpython#4938</li><li>python/cpython#5865</li><li>python/cpython#5871</li><li>python/cpython#6474</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/tiran'
closed_at = <Date 2018-02-25.09:19:20.791>
created_at = <Date 2017-11-30.18:05:09.552>
labels = ['expert-SSL', '3.8', 'type-bug', '3.7']
title = 'SSLContext.wrap_socket sends SNI Extension when server_hostname is IP'
updated_at = <Date 2018-04-14.22:19:00.613>
user = 'https://github.com/nitzmahone'
```

bugs.python.org fields:
```python
activity = <Date 2018-04-14.22:19:00.613>
actor = 'steve.dower'
assignee = 'christian.heimes'
closed = True
closed_date = <Date 2018-02-25.09:19:20.791>
closer = 'christian.heimes'
components = ['SSL']
creation = <Date 2017-11-30.18:05:09.552>
creator = 'nitzmahone'
dependencies = []
files = []
hgrepos = []
issue_num = 32185
keywords = ['patch', '3.5regression']
message_count = 9.0
messages = ['307333', '307334', '308712', '308713', '308715', '308716', '312782', '312786', '312787']
nosy_count = 10.0
nosy_names = ['paul.moore', 'janssen', 'pitrou', 'christian.heimes', 'tim.golden', 'alex', 'zach.ware', 'steve.dower', 'dstufft', 'nitzmahone']
pr_nums = ['4938', '5865', '5871', '6474']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue32185'
versions = ['Python 2.7', 'Python 3.6', 'Python 3.7', 'Python 3.8']
```

</p></details>

**Comment by efccb2b9-c9f3-44b4-8086-08881f25a009** [comment]
The current implementation of SSLContext.wrap_socket blindly sends whatever is passed in server_hostname in the SNI extension, assuming it's a DNS hostname. RFC6066 describes the SNI TLS extension, and specifically states that 'Literal IPv4 and IPv6 addresses are not permitted in "HostName".' The RFC makes no recommendation on how a server implementation that violates this requirement should behave; Microsoft's kernel HTTP listener (http.sys) chooses to abort the connection if SNI has been enabled. In the http.sys case, SNI is a global setting, currently off by default, but if any registered listener has SNI enabled, the connection abort behavior applies to all listeners.

SSLContext.wrap_socket() should determine whether server_hostname is an IP address before including the SNI extension. 

I've submitted a PR to work around this issue in urllib3 (https://github.com/shazow/urllib3/pull/1287) in the meantime, but would be good to get this fixed, especially if Microsoft decides to enable SNI by default at some point.

**Comment by tiran** [comment]
Thanks!

3.4 and 3.5 are out of scope. They only receive security fixes.

For 3.7 https://github.com/python/cpython/compare/master...tiran:openssl_check_hostname will take care of the issue

2.7 and 3.6 are a bit tricky. There is no platform-compatible way to detect if a string is an IP address. inet_pton() is not available on Windows. I cannot use the OpenSSL parser because it is only available in 1.0.2+. 2.7 and 3.6 still support 0.9.8.
